enterprise/api_client/discovery.py

```diff
@@ -155,6 +152,38 @@ def get_catalog_results(self, content_filter_query, query_params=None, traverse_
 
         return response
 
+    def get_course_id(self, course_identifier):
+        """
+        Return the course id for the given course identifier.  The `course_identifier` may be a course id or a course
+        run id; in either case the course id will be returned.
+
+        The 'course id' is the identifier for a course (ex. edX+DemoX)
+        The 'course run id' is the identifier for a run of a course (ex. edX+DemoX+demo_run)
+
+        Arguments:
+            course_identifier (str): The course id or course run id
+
+        Returns:
+            (str): course id
+        """
+
+        try:
+            CourseKey.from_string(course_identifier)
+        except InvalidKeyError:
+            # An `InvalidKeyError` is thrown if `course_identifier` is not in the proper format for a course run id.
+            # Since `course_identifier` is not a course run id we assume `course_identifier` is the  course id.
+            return course_identifier
+
+        # If here, `course_identifier` must be a course run id.
+        # We cannot use `CourseKey.from_string` to find the course id because that method assumes the course id is
+        # always a substring of the course run id and this is not always the case.  The only reliable way to determine
+        # which courses are associated with a given course run id is by by calling the discovery service.
+        course_run_data = self.get_course_run(course_identifier)
+        if 'course' in course_run_data:
+            return course_run_data['course']
+
+        return None
```

Comment:
`parse_course_key` only ever returned a string. So even though `None` makes sense in the abstract, does this bite us in any of the downstream calls of `get_course_id()`, since it's passing a false-y value instead of simply an incorrect string? I tried to follow some of the code paths and it seemed okay.. Just wanted to call it out since `get_course_id` isn't just a corrected version of `parse_course_key`.

Reply:
yeah that is valid.  I just wasn't sure what to do in the case that course_run_data did not have 'course' defined.  I chose `None` over an empty string because i figured returning `None` would fail louder and be easier to track down those bugs.  But I am open to other suggestions